### PR TITLE
Metric rewrite

### PR DESCRIPTION
**Pull request by cle-ros**
Changed most of the wrapper logic around metric computation. 

- extra_training_metrics can be passed to trainer
- extra_training_metrics and extra_validation_metrics can both be list, dict or None
- the metric interface has been changed to also have the aggregate function
- the train._aggregate_preds function is deprecated, because the aggregation is defined by the metric
- includes the merging of the "compute_batch" instead of compute_loss method in trainer
- metrics can have names that override the behavior when doing str(metric)
- added unit tests for metric aggregation
- modified integration test to include training metrics, list of extra metrics, and dict of extra metrics

### CODE REVIEW

File: flambe/learn/train.py [hunk 1]
```diff
@@ -45,7 +46,9 @@ def __init__(self,
                  lower_is_better: bool = False,
                  max_grad_norm: Optional[float] = None,
                  max_grad_abs_val: Optional[float] = None,
-                 extra_validation_metrics: Optional[Dict[str, Metric]] = None) -> None:
+                 extra_validation_metrics: Optional[Union[Dict[str, Metric], Iterable[Metric]]] = None,
+                 extra_training_metrics: Optional[Union[Dict[str, Metric], Iterable[Metric]]] = None,
+                 extra_training_metrics_log_interval: int = -1) -> None:
```
Comment (jeremyasapp) [inline flambe/learn/train.py]: doesn't need to be just for extras. Maybe let's treat the log interval for training metrics AND loss?

Comment (jeremyasapp) [reply flambe/learn/train.py]: -1 is also a weird default. Please use None.

Comment (jeremyasapp) [reply flambe/learn/train.py]: or 1 lol since that is currently the actual default

File: flambe/learn/train.py [hunk 2]
```diff
@@ -172,22 +218,66 @@ def _batch_to_device(self, batch: Tuple[torch.Tensor, ...]) -> Tuple[torch.Tenso
...
+        Parameters
+        ----------
+        log_prefix: str
+            A string, such as a tensorboard prefix
+        metrics_with_states: List[Tuple[Metric, Dict]]
+            a list of metric-state tuples
+        global_step: int
+            the global step for loggin
+
+        Returns
```
Comment (jeremyasapp) [inline flambe/learn/train.py]: remove if there is no return statement

File: flambe/learn/train.py [hunk 3]
```diff
@@ -172,22 +218,66 @@ def _batch_to_device(self, batch: Tuple[torch.Tensor, ...]) -> Tuple[torch.Tenso
...
+        batch = self._batch_to_device(batch)
+        pred, target = self.model(*batch)
+        for metric, state in metrics:
+            metric.aggregate(state, pred, target)
+        loss = self.loss_fn(pred, target)
+        return pred, target, loss
+
+    @staticmethod
+    def _log_metrics(log_prefix: str, metrics_with_states: List[Tuple[Metric, Dict]], global_step: int):
+        """
```
Comment (jeremyasapp) [inline flambe/learn/train.py]: wrong docstring format.

Don't jump a line at the start. Should always be:

"""One line description.

More info...

"""

Also keep it under 72 characters

File: flambe/learn/train.py [hunk 4]
```diff
@@ -229,14 +319,21 @@ def _train_step(self) -> None:
...
 
-            # Zero the gradients when exiting a train step
+                # Zero the gradients when exiting a train step
                 self.optimizer.zero_grad()
+                # logging train metrics
+                if self.extra_training_metrics_log_interval > self._last_train_log_step:
+                    self._log_metrics(log_prefix, metrics_with_states, global_step)
+                    self._last_train_log_step = i
+            if self._last_train_log_step != i:
+                # log again at end of epoch, if not logged at the end of epoch before
```
Comment (jeremyasapp) [inline flambe/learn/train.py]: is what you are calling an epoch here actually an epoch? Looks like it's at the end of a step rather, which can be any arbitrary number of iterations, as set by ``iter_per_step`` or ``epoch_per_step``

File: flambe/learn/train.py [hunk 5]
```diff
@@ -245,33 +342,40 @@ def _aggregate_preds(self, data_iterator: Iterator) -> Tuple[torch.Tensor, torch
 
         Returns
         -------
-        Tuple[torch.tensor, torch.tensor]
-            The predictions and targets.
+        Tuple[torch.tensor, torch.tensor, float]
+            The predictions, targets and mean loss.
+
+        DEPRECATED; only existed to aggregate for the metric functions. The metric functions do this in-place now.
```
Comment (jeremyasapp) [inline flambe/learn/train.py]: way too long. < 72 characters

File: flambe/metric/metric.py [hunk 6]
```diff
@@ -12,6 +13,18 @@ class Metric(Component):
     examples and provide as output a processd list of the same size.
 
     """
+    def __init__(self, name: Optional[str] = None):
```
Comment (jeremyasapp) [inline flambe/metric/metric.py]: giving a name to all metrics, while also allowing the input to trainer to be a dict seems redundant. I guess I understand now why you wanted to make it a list. Designwise I find this solution a bit akward. Why would a metric needs to know its own name you know? Any reason not to keep the status quo? No names in the Metric and Dict only as input? I don't really get the benefits that this change provides

File: flambe/metric/metric.py [hunk 7]
```diff
@@ -32,10 +45,58 @@ def compute(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
...
+        return state['accumulated_score'] if 'accumulated_score' in state else None
+
     def __call__(self, *args, **kwargs):
         """Makes Featurizer a callable."""
         return self.compute(*args, **kwargs)
 
     def __str__(self) -> str:
         """Return the name of the Metric (for use in logging)."""
-        return self.__class__.__name__
+        return self.__class__.__name__ if getattr(self, 'name', None) is None else self.name
```
Comment (jeremyasapp) [inline flambe/metric/metric.py]: calling getattr is weird here. Just set self.name to None in the parent __init__ and use self.name is None here

File: flambe/optim/noam.py [hunk 8]
```diff
@@ -53,4 +53,4 @@ def lr_lambda(self, step: int) -> float:
             if step > self.warmup:
                 return 1. / (self.d_model ** 0.5) / (step ** 0.5)
             else:
-                return step / (self.d_model ** 0.5) / (self.warmup ** 1.5)
+                return (step + 1) / (self.d_model ** 0.5) / (self.warmup ** 1.5)
```
Comment (jeremyasapp) [inline flambe/optim/noam.py]: This is not part of the canonical implementation by the original authors. I don't think we should change that for consistency.

File: tests/unit/metrics/test_dev.py [hunk 9]
```diff
@@ -58,6 +58,27 @@ def test_accuracy():
...
+    metric_state = {}
+    metric = Accuracy()
+    metric.aggregate(metric_state, torch.tensor([[0.1, 0.2], [0.9, 0.1]]), torch.tensor([1, 1]))
+    assert metric_state['accumulated_score'] == 0.5
+    metric.aggregate(metric_state, torch.tensor([[1.0, 0.0], [0.6, 0.4]]), torch.tensor([1, 1]))
+    assert metric_state['accumulated_score'] == 0.25
+    assert metric_state['sample_count'] == 4
+    assert metric.finalize(metric_state) == 0.25
+
+    # for AUC (different aggregation functionality)
```
Comment (jeremyasapp) [inline tests/unit/metrics/test_dev.py]: please separate this test into 2. One for Accuracy, and one for perplexity

File: .gitignore [hunk 10]
```diff
@@ -277,4 +277,6 @@ tags
 # Ignore all local history of files
 .history
 
+### for local testing
```
Comment (jeremyasapp) [inline .gitignore]: what's that about?

File: flambe/metric/metric.py [hunk 11]
```diff
@@ -32,6 +34,55 @@ def compute(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
...
+        args:
+            normally pred, target
+        kwargs
+
+        Returns
+        -------
+        The updated state (even though the update happens in-place)
+        """
+        score = self.compute(*args, **kwargs)
+        if isinstance(score, torch.Tensor):
```
Comment (brendandc-asapp) [inline flambe/metric/metric.py]: nit: a ternary reads cleaner here!

File: flambe/metric/metric.py [hunk 12]
```diff
@@ -32,6 +34,55 @@ def compute(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
...
+            score_np = score.cpu().detach().numpy()
+        else:
+            score_np = score
+        try:
+            num_samples = args[0].size(0)
+        except (ValueError, AttributeError):
+            raise ValueError(f'Cannot get size from {type(args[0])}')
+        if state == {}:
```
Comment (brendandc-asapp) [inline flambe/metric/metric.py]: ditto on the defaultdict comment above, and simplifying the declaration if not

File: flambe/metric/metric.py [hunk 13]
```diff
@@ -32,6 +34,55 @@ def compute(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
...
+        Parameters
+        ----------
+        state: dict
+            the metric state
+
+        Returns
+        -------
+        The final score
+        """
+        return state['accumulated_score'] if 'accumulated_score' in state else None
```
Comment (brendandc-asapp) [inline flambe/metric/metric.py]:
```suggestion
        return state.get('accumulated_score')
```

File: flambe/metric/dev/auc.py [hunk 14]
```diff
@@ -18,6 +18,54 @@ def __init__(self, max_fpr=1.0):
...
+        state: dict
+            the metric state
+        args: the pred, target tuple
+
+        Returns
+        -------
+
+        """
+        pred, target = args
+        if state == {}:
```
[the remark this answers is not in the document]
Comment (cle-ros) [reply flambe/metric/dev/auc.py]: no. modification has to happen in-place

File: flambe/metric/metric.py (hunk 12 above)
Comment (cle-ros) [reply flambe/metric/metric.py]: as above,  modification has to happen in-place

File: flambe/metric/loss/cross_entropy.py [hunk 15]
```diff
@@ -11,7 +11,7 @@ class MultiLabelCrossEntropy(Metric):
     def __init__(self,
                  weight: Optional[torch.Tensor] = None,
                  ignore_index: Optional[int] = None,
-                 reduction: str = 'mean') -> None:
+                 reduction: str = 'mean',) -> None:
```
Comment (jeremyasapp) [inline flambe/metric/loss/cross_entropy.py]: why this comma?

File: flambe/metric/metric.py (hunk 12 above)
Comment (jeremyasapp) [reply flambe/metric/metric.py]: you can check emptyness with `if not state`

File: flambe/metric/dev/perplexity.py [hunk 16]
```diff
@@ -7,7 +7,9 @@ class Perplexity(Metric):
     """Token level perplexity, computed a exp(cross_entropy)."""
 
     def __init__(self):
-        """Initalizes the Perplexity metrc."""
+        """
```
Comment (jeremyasapp) [inline flambe/metric/dev/perplexity.py]: wrong docstring format

File: flambe/metric/dev/bpc.py [hunk 17]
```diff
@@ -7,7 +7,9 @@ class BPC(Metric):
     """Bits per character. Computed as log_2(perplexity)"""
 
     def __init__(self):
-        """Initalizes the Metric."""
+        """
```
Comment (jeremyasapp) [inline flambe/metric/dev/bpc.py]: Wrong docstring format

File: flambe/metric/dev/auc.py [hunk 18]
```diff
@@ -18,6 +21,54 @@ def __init__(self, max_fpr=1.0):
...
+            state['pred'] = []
+            state['target'] = []
+        state['pred'].append(pred.cpu().detach())
+        state['target'].append(target.cpu().detach())
+        return state
+
+    def finalize(self, state) -> float:
+        """
+        FInalizes the metric computation
```
Comment (jeremyasapp) [inline flambe/metric/dev/auc.py]: typo and wrong docstring format

File: flambe/metric/dev/auc.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (jeremyasapp) [reply flambe/metric/dev/auc.py]: should do `if not state`
